File: app/modules/blackout-signal/android/src/main/java/expo/modules/blackoutsignal/BlackoutSignalModule.kt
package expo.modules.blackoutsignal

// ------------------------------------------------------------------
// Pont Expo -> libsignal officiel (org.signal:libsignal-android).
// Implemente le contrat SignalBridge de l'app (signalBridge.types.ts).
//
// Modele "coeur fonctionnel" : AUCUN etat conserve ici. Chaque appel
// recoit les records serialises (base64), reconstruit des stores
// in-memory libsignal le temps de l'operation, et retourne les
// records mis a jour. La persistance appartient a la couche JS
// (SQLCipher). Les cles privees ne quittent jamais l'appareil : les
// records ne transitent que sur le pont JS<->natif local.
//
// La confiance dans les identites n'est PAS decidee ici (les stores
// in-memory acceptent tout) : le pinning est fait par la couche app
// via la verification QR out-of-band.
// ------------------------------------------------------------------

import android.util.Base64
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import java.security.SecureRandom
import org.signal.libsignal.protocol.IdentityKey
import org.signal.libsignal.protocol.IdentityKeyPair
import org.signal.libsignal.protocol.SessionBuilder
import org.signal.libsignal.protocol.SessionCipher
import org.signal.libsignal.protocol.SignalProtocolAddress
import org.signal.libsignal.protocol.ecc.ECKeyPair
import org.signal.libsignal.protocol.ecc.ECPublicKey
import org.signal.libsignal.protocol.kem.KEMKeyPair
import org.signal.libsignal.protocol.kem.KEMKeyType
import org.signal.libsignal.protocol.kem.KEMPublicKey
import org.signal.libsignal.protocol.message.PreKeySignalMessage
import org.signal.libsignal.protocol.message.SignalMessage
import org.signal.libsignal.protocol.state.KyberPreKeyRecord
import org.signal.libsignal.protocol.state.PreKeyBundle
import org.signal.libsignal.protocol.state.PreKeyRecord
import org.signal.libsignal.protocol.state.SessionRecord
import org.signal.libsignal.protocol.state.SignedPreKeyRecord
import org.signal.libsignal.protocol.state.impl.InMemorySignalProtocolStore

class BundleParam : Record {
  @Field var registrationId: Int = 0
  @Field var deviceId: Int = 1
  @Field var identityKey: String = ""
  @Field var signedPreKeyId: Int = 0
  @Field var signedPreKeyPublic: String = ""
  @Field var signedPreKeySignature: String = ""
  @Field var preKeyId: Int? = null
  @Field var preKeyPublic: String? = null
  @Field var kyberPreKeyId: Int = 0
  @Field var kyberPreKeyPublic: String = ""
  @Field var kyberPreKeySignature: String = ""
}

class LocalPreKeysParam : Record {
  @Field var signedPreKeyRecord: String = ""
  @Field var kyberPreKeyRecord: String = ""
  @Field var preKeyRecords: List<String> = emptyList()
}

private fun b64(bytes: ByteArray): String = Base64.encodeToString(bytes, Base64.NO_WRAP)
private fun unb64(s: String): ByteArray = Base64.decode(s, Base64.NO_WRAP)

class BlackoutSignalModule : Module() {

  private fun makeStore(identityRecord: String, registrationId: Int): InMemorySignalProtocolStore {
    val identity = IdentityKeyPair(unb64(identityRecord))
    return InMemorySignalProtocolStore(identity, registrationId)
  }

  override fun definition() = ModuleDefinition {
    Name("BlackoutSignal")

    AsyncFunction("generateIdentityKeyPair") {
      val keyPair = IdentityKeyPair.generate()
      // 14 bits non nuls, meme convention que KeyHelper de libsignal
      val registrationId = SecureRandom().nextInt(16380) + 1
      mapOf(
        "identityRecord" to b64(keyPair.serialize()),
        "publicKey" to b64(keyPair.publicKey.serialize()),
        "registrationId" to registrationId,
      )
    }

    AsyncFunction("generatePreKeys") {
        identityRecord: String, signedPreKeyId: Int, preKeyStartId: Int, preKeyCount: Int, kyberPreKeyId: Int ->
      val identity = IdentityKeyPair(unb64(identityRecord))
      val now = System.currentTimeMillis()

      val spkPair = ECKeyPair.generate()
      val spkSignature = identity.privateKey.calculateSignature(spkPair.publicKey.serialize())
      val signedPreKey = SignedPreKeyRecord(signedPreKeyId, now, spkPair, spkSignature)

      val preKeys = (0 until preKeyCount).map { i ->
        val pair = ECKeyPair.generate()
        val record = PreKeyRecord(preKeyStartId + i, pair)
        mapOf(
          "id" to preKeyStartId + i,
          "record" to b64(record.serialize()),
          "publicKey" to b64(pair.publicKey.serialize()),
        )
      }

      val kemPair = KEMKeyPair.generate(KEMKeyType.KYBER_1024)
      val kyberSignature = identity.privateKey.calculateSignature(kemPair.publicKey.serialize())
      val kyberPreKey = KyberPreKeyRecord(kyberPreKeyId, now, kemPair, kyberSignature)

      mapOf(
        "signedPreKey" to mapOf(
          "id" to signedPreKeyId,
          "record" to b64(signedPreKey.serialize()),
          "publicKey" to b64(spkPair.publicKey.serialize()),
          "signature" to b64(spkSignature),
        ),
        "preKeys" to preKeys,
        "kyberPreKey" to mapOf(
          "id" to kyberPreKeyId,
          "record" to b64(kyberPreKey.serialize()),
          "publicKey" to b64(kemPair.publicKey.serialize()),
          "signature" to b64(kyberSignature),
        ),
      )
    }

    AsyncFunction("processPreKeyBundle") {
        identityRecord: String, registrationId: Int, localAddress: String, remoteAddress: String, bundle: BundleParam ->
      val store = makeStore(identityRecord, registrationId)
      val remote = SignalProtocolAddress(remoteAddress, bundle.deviceId)
      val local = SignalProtocolAddress(localAddress, 1)

      val preKeyBundle = PreKeyBundle(
        bundle.registrationId,
        bundle.deviceId,
        bundle.preKeyId ?: PreKeyBundle.NULL_PRE_KEY_ID,
        bundle.preKeyPublic?.let { ECPublicKey(unb64(it)) },
        bundle.signedPreKeyId,
        ECPublicKey(unb64(bundle.signedPreKeyPublic)),
        unb64(bundle.signedPreKeySignature),
        IdentityKey(unb64(bundle.identityKey)),
        bundle.kyberPreKeyId,
        KEMPublicKey(unb64(bundle.kyberPreKeyPublic)),
        unb64(bundle.kyberPreKeySignature),
      )

      SessionBuilder(store, remote, local).process(preKeyBundle)
      mapOf("session" to b64(store.loadSession(remote).serialize()))
    }

    AsyncFunction("encrypt") {
        identityRecord: String, registrationId: Int, localAddress: String, remoteAddress: String, session: String, plaintext: String ->
      val store = makeStore(identityRecord, registrationId)
      val remote = SignalProtocolAddress(remoteAddress, 1)
      val local = SignalProtocolAddress(localAddress, 1)
      store.storeSession(remote, SessionRecord(unb64(session)))

      val message = SessionCipher(store, local, remote).encrypt(unb64(plaintext))
      mapOf(
        "type" to message.type,
        "body" to b64(message.serialize()),
        "session" to b64(store.loadSession(remote).serialize()),
      )
    }

    AsyncFunction("decryptWhisper") {
        identityRecord: String, registrationId: Int, localAddress: String, remoteAddress: String, session: String, body: String ->
      val store = makeStore(identityRecord, registrationId)
      val remote = SignalProtocolAddress(remoteAddress, 1)
      val local = SignalProtocolAddress(localAddress, 1)
      store.storeSession(remote, SessionRecord(unb64(session)))

      val plaintext = SessionCipher(store, local, remote).decrypt(SignalMessage(unb64(body)))
      mapOf(
        "plaintext" to b64(plaintext),
        "session" to b64(store.loadSession(remote).serialize()),
      )
    }

    AsyncFunction("decryptPreKey") {
        identityRecord: String, registrationId: Int, localAddress: String, remoteAddress: String,
        session: String?, body: String, localPreKeys: LocalPreKeysParam ->
      val store = makeStore(identityRecord, registrationId)
      val remote = SignalProtocolAddress(remoteAddress, 1)
      val local = SignalProtocolAddress(localAddress, 1)
      if (session != null) store.storeSession(remote, SessionRecord(unb64(session)))

      val spk = SignedPreKeyRecord(unb64(localPreKeys.signedPreKeyRecord))
      store.storeSignedPreKey(spk.id, spk)
      val kyber = KyberPreKeyRecord(unb64(localPreKeys.kyberPreKeyRecord))
      store.storeKyberPreKey(kyber.id, kyber)
      for (recordB64 in localPreKeys.preKeyRecords) {
        val record = PreKeyRecord(unb64(recordB64))
        store.storePreKey(record.id, record)
      }

      val message = PreKeySignalMessage(unb64(body))
      val plaintext = SessionCipher(store, local, remote).decrypt(message)
      mapOf(
        "plaintext" to b64(plaintext),
        "session" to b64(store.loadSession(remote).serialize()),
        "usedPreKeyId" to if (message.preKeyId.isPresent) message.preKeyId.get() else null,
      )
    }
  }
}

File: app/modules/blackout-signal/ios/BlackoutSignalModule.swift
import ExpoModulesCore
import LibSignalClient

// ------------------------------------------------------------------
// Pont Expo -> LibSignalClient (lib Swift officielle de Signal).
// Implemente le contrat SignalBridge de l'app (signalBridge.types.ts),
// strictement equivalent au module Kotlin et a l'adaptateur Node des
// tests (meme coeur Rust dans les trois cas).
//
// Modele "coeur fonctionnel" : aucun etat conserve ici. Records
// serialises (base64) en entree, records mis a jour en sortie ;
// persistance cote JS (SQLCipher). Les stores in-memory acceptent
// toute identite : le pinning est fait par la couche app (QR).
// ------------------------------------------------------------------

struct BundleParam: Record {
  @Field var registrationId: Int = 0
  @Field var deviceId: Int = 1
  @Field var identityKey: String = ""
  @Field var signedPreKeyId: Int = 0
  @Field var signedPreKeyPublic: String = ""
  @Field var signedPreKeySignature: String = ""
  @Field var preKeyId: Int? = nil
  @Field var preKeyPublic: String? = nil
  @Field var kyberPreKeyId: Int = 0
  @Field var kyberPreKeyPublic: String = ""
  @Field var kyberPreKeySignature: String = ""
}

struct LocalPreKeysParam: Record {
  @Field var signedPreKeyRecord: String = ""
  @Field var kyberPreKeyRecord: String = ""
  @Field var preKeyRecords: [String] = []
}

enum BlackoutSignalError: Error {
  case badBase64(String)
}

private func unb64(_ s: String) throws -> Data {
  guard let data = Data(base64Encoded: s) else { throw BlackoutSignalError.badBase64(s) }
  return data
}

private func b64(_ d: Data) -> String { d.base64EncodedString() }

private func makeStore(_ identityRecord: String, _ registrationId: Int) throws -> InMemorySignalProtocolStore {
  let identity = try IdentityKeyPair(bytes: unb64(identityRecord))
  return InMemorySignalProtocolStore(identity: identity, registrationId: UInt32(registrationId))
}

private func loadSessionB64(_ store: InMemorySignalProtocolStore, _ remote: ProtocolAddress) throws -> String {
  guard let session = try store.loadSession(for: remote, context: NullContext()) else {
    throw SignalError.sessionNotFound("session absente apres l'operation")
  }
  return b64(session.serialize())
}

public class BlackoutSignalModule: Module {
  public func definition() -> ModuleDefinition {
    Name("BlackoutSignal")

    AsyncFunction("generateIdentityKeyPair") { () -> [String: Any] in
      let keyPair = IdentityKeyPair.generate()
      // 14 bits non nuls, meme convention que KeyHelper de libsignal
      let registrationId = Int.random(in: 1...16380)
      return [
        "identityRecord": b64(keyPair.serialize()),
        "publicKey": b64(keyPair.publicKey.serialize()),
        "registrationId": registrationId,
      ]
    }

    AsyncFunction("generatePreKeys") {
      (identityRecord: String, signedPreKeyId: Int, preKeyStartId: Int, preKeyCount: Int, kyberPreKeyId: Int) -> [String: Any] in
      let identity = try IdentityKeyPair(bytes: unb64(identityRecord))
      let now = UInt64(Date().timeIntervalSince1970 * 1000)

      let spkPrivate = PrivateKey.generate()
      let spkSignature = identity.privateKey.generateSignature(message: spkPrivate.publicKey.serialize())
      let signedPreKey = try SignedPreKeyRecord(
        id: UInt32(signedPreKeyId), timestamp: now, privateKey: spkPrivate, signature: spkSignature)

      var preKeys: [[String: Any]] = []
      for i in 0..<preKeyCount {
        let priv = PrivateKey.generate()
        let record = try PreKeyRecord(id: UInt32(preKeyStartId + i), privateKey: priv)
        preKeys.append([
          "id": preKeyStartId + i,
          "record": b64(record.serialize()),
          "publicKey": b64(priv.publicKey.serialize()),
        ])
      }

      let kemPair = KEMKeyPair.generate()
      let kyberSignature = identity.privateKey.generateSignature(message: kemPair.publicKey.serialize())
      let kyberPreKey = try KyberPreKeyRecord(
        id: UInt32(kyberPreKeyId), timestamp: now, keyPair: kemPair, signature: kyberSignature)

      return [
        "signedPreKey": [
          "id": signedPreKeyId,
          "record": b64(signedPreKey.serialize()),
          "publicKey": b64(spkPrivate.publicKey.serialize()),
          "signature": b64(spkSignature),
        ],
        "preKeys": preKeys,
        "kyberPreKey": [
          "id": kyberPreKeyId,
          "record": b64(kyberPreKey.serialize()),
          "publicKey": b64(kemPair.publicKey.serialize()),
          "signature": b64(kyberSignature),
        ],
      ]
    }

    AsyncFunction("processPreKeyBundle") {
      (identityRecord: String, registrationId: Int, localAddress: String, remoteAddress: String, bundle: BundleParam) -> [String: Any] in
      let store = try makeStore(identityRecord, registrationId)
      let remote = try ProtocolAddress(name: remoteAddress, deviceId: UInt32(bundle.deviceId))
      let local = try ProtocolAddress(name: localAddress, deviceId: 1)

      let identityKey = try IdentityKey(bytes: unb64(bundle.identityKey))
      let signedPrekey = try PublicKey(unb64(bundle.signedPreKeyPublic))
      let kyberPrekey = try KEMPublicKey(unb64(bundle.kyberPreKeyPublic))

      let preKeyBundle: PreKeyBundle
      if let preKeyId = bundle.preKeyId, let preKeyPublic = bundle.preKeyPublic {
        preKeyBundle = try PreKeyBundle(
          registrationId: UInt32(bundle.registrationId),
          deviceId: UInt32(bundle.deviceId),
          prekeyId: UInt32(preKeyId),
          prekey: try PublicKey(unb64(preKeyPublic)),
          signedPrekeyId: UInt32(bundle.signedPreKeyId),
          signedPrekey: signedPrekey,
          signedPrekeySignature: try unb64(bundle.signedPreKeySignature),
          identity: identityKey,
          kyberPrekeyId: UInt32(bundle.kyberPreKeyId),
          kyberPrekey: kyberPrekey,
          kyberPrekeySignature: try unb64(bundle.kyberPreKeySignature))
      } else {
        preKeyBundle = try PreKeyBundle(
          registrationId: UInt32(bundle.registrationId),
          deviceId: UInt32(bundle.deviceId),
          signedPrekeyId: UInt32(bundle.signedPreKeyId),
          signedPrekey: signedPrekey,
          signedPrekeySignature: try unb64(bundle.signedPreKeySignature),
          identity: identityKey,
          kyberPrekeyId: UInt32(bundle.kyberPreKeyId),
          kyberPrekey: kyberPrekey,
          kyberPrekeySignature: try unb64(bundle.kyberPreKeySignature))
      }

      try processPreKeyBundle(
        preKeyBundle, for: remote, ourAddress: local,
        sessionStore: store, identityStore: store, context: NullContext())

      return ["session": try loadSessionB64(store, remote)]
    }

    AsyncFunction("encrypt") {
      (identityRecord: String, registrationId: Int, localAddress: String, remoteAddress: String, session: String, plaintext: String) -> [String: Any] in
      let store = try makeStore(identityRecord, registrationId)
      let remote = try ProtocolAddress(name: remoteAddress, deviceId: 1)
      let local = try ProtocolAddress(name: localAddress, deviceId: 1)
      try store.storeSession(SessionRecord(bytes: unb64(session)), for: remote, context: NullContext())

      let message = try signalEncrypt(
        message: unb64(plaintext), for: remote, localAddress: local,
        sessionStore: store, identityStore: store, context: NullContext())

      return [
        "type": Int(message.messageType.rawValue),
        "body": b64(message.serialize()),
        "session": try loadSessionB64(store, remote),
      ]
    }

    AsyncFunction("decryptWhisper") {
      (identityRecord: String, registrationId: Int, localAddress: String, remoteAddress: String, session: String, body: String) -> [String: Any] in
      let store = try makeStore(identityRecord, registrationId)
      let remote = try ProtocolAddress(name: remoteAddress, deviceId: 1)
      let local = try ProtocolAddress(name: localAddress, deviceId: 1)
      try store.storeSession(SessionRecord(bytes: unb64(session)), for: remote, context: NullContext())

      let message = try SignalMessage(bytes: unb64(body))
      let plaintext = try signalDecrypt(
        message: message, from: remote, to: local,
        sessionStore: store, identityStore: store, context: NullContext())

      return [
        "plaintext": b64(plaintext),
        "session": try loadSessionB64(store, remote),
      ]
    }

    AsyncFunction("decryptPreKey") {
      (identityRecord: String, registrationId: Int, localAddress: String, remoteAddress: String,
       session: String?, body: String, localPreKeys: LocalPreKeysParam) -> [String: Any] in
      let store = try makeStore(identityRecord, registrationId)
      let remote = try ProtocolAddress(name: remoteAddress, deviceId: 1)
      let local = try ProtocolAddress(name: localAddress, deviceId: 1)
      if let session = session {
        try store.storeSession(SessionRecord(bytes: unb64(session)), for: remote, context: NullContext())
      }

      let spk = try SignedPreKeyRecord(bytes: unb64(localPreKeys.signedPreKeyRecord))
      try store.storeSignedPreKey(spk, id: spk.id, context: NullContext())
      let kyber = try KyberPreKeyRecord(bytes: unb64(localPreKeys.kyberPreKeyRecord))
      try store.storeKyberPreKey(kyber, id: kyber.id, context: NullContext())
      for recordB64 in localPreKeys.preKeyRecords {
        let record = try PreKeyRecord(bytes: unb64(recordB64))
        try store.storePreKey(record, id: record.id, context: NullContext())
      }

      let message = try PreKeySignalMessage(bytes: unb64(body))
      let plaintext = try signalDecryptPreKey(
        message: message, from: remote, localAddress: local,
        sessionStore: store, identityStore: store,
        preKeyStore: store, signedPreKeyStore: store, kyberPreKeyStore: store,
        context: NullContext())

      return [
        "plaintext": b64(plaintext),
        "session": try loadSessionB64(store, remote),
        "usedPreKeyId": try message.preKeyId().map { Int($0) } as Any,
      ]
    }
  }
}
